### PR TITLE
[Core] Add geometry metric calculation utilities

### PR DESCRIPTION
**Pull request by rubenzorrilla**
**Description**
Addition of utilities to calculate the metric tensor of a given geometry. So far this only considers triangles and tetras but I implemented it in such a way it is extensible to other geometry types. Unfortunately, for more "involved" geometries the metric calculation is way more complex.

The new methods allow calculating only the metric tensor or the metric tensor with some extra data. Such data requires solving the eigenvalues of the metric tensor so I think is good to keep them separated to avoid such operation in case it is not needed. Besides this I also added one that returns the dimensionless metric.

Tests have been added accordingly.

Merging this utility is required in order to separate the physics-based shock capturing from the compressible NS RK4 strategy #7831 .

**Changelog**
- Metric tensor calculation utilities addition

### CODE REVIEW

File: kratos/utilities/geometry_metric_calculator.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    ///@{
+
+    // Deleted default constructor
+    GeometryMetricCalculator() = delete;
+
+    /// Deleted copy constructor.
+    GeometryMetricCalculator(GeometryMetricCalculator const& rOther) = delete;
+
+    /// Destructor.
+    ~GeometryMetricCalculator();
```
Comment (philbucher) [inline kratos/utilities/geometry_metric_calculator.h]:
```suggestion
```
better delete it, also the cpp implementaiton is missing

**Comment by philbucher** [comment]
FYI, maybe you know this already: https://github.com/KratosMultiphysics/Kratos/wiki/KRATOS_API-for-Windows-Compilation

**Comment by rubenzorrilla** [comment]
> FYI, maybe you know this already: https://github.com/KratosMultiphysics/Kratos/wiki/KRATOS_API-for-Windows-Compilation

The point is that the `KRATOS_API` is correct in principle (there is no problem in my machine)... It is most probably a stupid mistake... :rage:

**Comment by philbucher** [comment]
:/

After you figured it out and it is missing in the documentation, can you please add it ? Thx :)

**Comment by rubenzorrilla** [comment]
> :/
> 
> After you figured it out and it is missing in the documentation, can you please add it ? Thx :)

The problem was that I copied and pasted the include ward (without modifying it) so the header wasn't included :sweat:

File: kratos/utilities/geometry_metric_calculator.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+     * @param rGeometry Reference to the geometry of interest
+     * @param rMetricTensor Reference to the metric tensor
+     * @param rReferenceElementSize Reference to the reference element size
+     * @param rMetricInfimum Reference to the metric infimum
+     * @param rMetricSupremum Reference to the metric supremum
+     */
+    static void CalculateMetricTensor(
+        const GeometryType& rGeometry,
+        BoundedMatrix<double, TDim, TDim>& rMetricTensor,
+        double& rReferenceElementSize,
```
Comment (RiccardoRossi) [inline kratos/utilities/geometry_metric_calculator.h:116]: this would look sooo much cooler with c++17 :-(

**Review by RiccardoRossi — DISMISSED** [review]

**Review by loumalouomega — COMMENTED** [review]
Just to say that I am Okay, but I prefer std::size_t instead of unsigned int

**Comment by rubenzorrilla** [comment]
> Just to say that I am Okay, but I prefer std::size_t instead of unsigned int

May I ask why? Is there any advantage? If so it is I don't mind changing it.

**Comment by loumalouomega** [comment]
> > Just to say that I am Okay, but I prefer std::size_t instead of unsigned int
> 
> May I ask why? Is there any advantage? If so it is I don't mind changing it.

The most important one... aesthetics and make me happy :D, usually we use indirectly with IndexType

**Comment by RiccardoRossi** [comment]
actually vicente is right on this one (even though i am more aesthetically pleased by unsigned int)

[link](https://stackoverflow.com/questions/131803/unsigned-int-vs-size-t
)

**Comment by RiccardoRossi** [comment]
for short loops it does not matter though

**Comment by rubenzorrilla** [comment]
I don't think it matters in this case. Nevertheless, and considering that @loumalouomega is a potential user of this, I changed it to make him happy :wink: .

**Comment by loumalouomega** [comment]
> I don't think it matters in this case. Nevertheless, and considering that @loumalouomega is a potential user of this, I changed it to make him happy wink .

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**